# Data Cleaning

## Data loading

Load the dataset from "sem_3_ret_modelling.csv" into a dataframe.


In [18]:
import pandas as pd

location = "/content/drive/MyDrive/projects/hemlc_second/data/"

df = pd.read_csv(f"{location}sem_3_ret_modelling.csv")
display(df.head())

SID     COHORT             RACE_ETHNICITY  GENDER  \
0  UHDOP5522  Fall 2020                      Asian  Female   
1  UHE842CU6  Fall 2021  Black or African American  Female   
2  UHJFT1JAB  Fall 2018                      Asian  Female   
3  UHKF05TAF  Fall 2018                   Hispanic  Female   
4  UHKKQ8UY5  Fall 2021                   Hispanic    Male   

        FIRST_GEN_STATUS FAMILY_INCOME  HS_MATH_GPA  HS_ENGL_GPA  \
0  Continuing Generation           NaN          3.2        3.400   
1  Continuing Generation           NaN          2.6        3.750   
2  Continuing Generation  100K to 149K          3.4        3.500   
3       First Generation           NaN          3.0        3.375   
4  Continuing Generation           NaN          2.5        2.625   

                    COLLEGE  UNITS_ATTEMPTED_1  UNITS_ATTEMPTED_2  \
0  Visual & Performing Arts               15.0               14.0   
1  Visual & Performing Arts               12.0               12.0   
2  Visual & Performing Arts               15.0               15.0   
3      Letters & Humanities               16.0                9.0   
4      Letters & Humanities               13.0               13.0   

   UNITS_COMPLETED_1  UNITS_COMPLETED_2     GPA_1     GPA_2  GPA_3  \
0               15.0               15.0  4.000000  3.785714    4.0   
1               12.0               12.0  3.000000  2.500000    1.5   
2               15.0               16.0  3.800000  3.600000    3.6   
3                7.0                3.0  1.562500  1.000000    2.5   
4               13.0               13.0  3.538462  3.769231    3.4   

  SEM_1_STATUS SEM_2_STATUS SEM_3_STATUS  
0            E            E            E  
1            E            E            E  
2            E            E            E  
3            E            E            E  
4            E            E            E

Display information about the DataFrame, including the data types and non-null values to understand the structure and identify missing values.



In [19]:
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25266 entries, 0 to 25265
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   SID                25266 non-null  object 
 1   COHORT             25266 non-null  object 
 2   RACE_ETHNICITY     25266 non-null  object 
 3   GENDER             25266 non-null  object 
 4   FIRST_GEN_STATUS   25266 non-null  object 
 5   FAMILY_INCOME      4295 non-null   object 
 6   HS_MATH_GPA        24906 non-null  float64
 7   HS_ENGL_GPA        24906 non-null  float64
 8   COLLEGE            25266 non-null  object 
 9   UNITS_ATTEMPTED_1  25136 non-null  float64
 10  UNITS_ATTEMPTED_2  24229 non-null  float64
 11  UNITS_COMPLETED_1  25266 non-null  float64
 12  UNITS_COMPLETED_2  24362 non-null  float64
 13  GPA_1              25136 non-null  float64
 14  GPA_2              24229 non-null  float64
 15  GPA_3              22198 non-null  float64
 16  SEM_1_STATUS       252

None

Check for the number of missing values in each column.



In [20]:
display(df.isnull().sum())

SID                      0
COHORT                   0
RACE_ETHNICITY           0
GENDER                   0
FIRST_GEN_STATUS         0
FAMILY_INCOME        20971
HS_MATH_GPA            360
HS_ENGL_GPA            360
COLLEGE                  0
UNITS_ATTEMPTED_1      130
UNITS_ATTEMPTED_2     1037
UNITS_COMPLETED_1        0
UNITS_COMPLETED_2      904
GPA_1                  130
GPA_2                 1037
GPA_3                 3068
SEM_1_STATUS             0
SEM_2_STATUS             0
SEM_3_STATUS             0
dtype: int64

Identify columns with more than 50% missing values and drop them from the dataframe.



Drop columns with excessive missing values based on the previous exploration.


In [21]:
missing_values_count = df.isnull().sum()
total_rows = len(df)
columns_to_drop = missing_values_count[missing_values_count / total_rows > 0.5].index.tolist()
df.drop(columns=columns_to_drop, inplace=True)
display(f"Number of remaining columns: {df.shape[1]}")
display(df.head())

'Number of remaining columns: 18'

SID     COHORT             RACE_ETHNICITY  GENDER  \
0  UHDOP5522  Fall 2020                      Asian  Female   
1  UHE842CU6  Fall 2021  Black or African American  Female   
2  UHJFT1JAB  Fall 2018                      Asian  Female   
3  UHKF05TAF  Fall 2018                   Hispanic  Female   
4  UHKKQ8UY5  Fall 2021                   Hispanic    Male   

        FIRST_GEN_STATUS  HS_MATH_GPA  HS_ENGL_GPA                   COLLEGE  \
0  Continuing Generation          3.2        3.400  Visual & Performing Arts   
1  Continuing Generation          2.6        3.750  Visual & Performing Arts   
2  Continuing Generation          3.4        3.500  Visual & Performing Arts   
3       First Generation          3.0        3.375      Letters & Humanities   
4  Continuing Generation          2.5        2.625      Letters & Humanities   

   UNITS_ATTEMPTED_1  UNITS_ATTEMPTED_2  UNITS_COMPLETED_1  UNITS_COMPLETED_2  \
0               15.0               14.0               15.0               15.0   
1               12.0               12.0               12.0               12.0   
2               15.0               15.0               15.0               16.0   
3               16.0                9.0                7.0                3.0   
4               13.0               13.0               13.0               13.0   

      GPA_1     GPA_2  GPA_3 SEM_1_STATUS SEM_2_STATUS SEM_3_STATUS  
0  4.000000  3.785714    4.0            E            E            E  
1  3.000000  2.500000    1.5            E            E            E  
2  3.800000  3.600000    3.6            E            E            E  
3  1.562500  1.000000    2.5            E            E            E  
4  3.538462  3.769231    3.4            E            E            E

Check for duplicate rows in the DataFrame.



In [22]:
display(df.duplicated().sum())

np.int64(68)

Drop duplicate rows from the DataFrame.



In [23]:
df.drop_duplicates(inplace=True)
display(df.duplicated().sum())

np.int64(0)

Inspect the unique values and their counts for the categorical columns to identify any anomalies or labels that need fixing.



In [24]:
categorical_cols = df.select_dtypes(include='object').columns
for col in categorical_cols:
    display(f"Value counts for column: {col}")
    display(df[col].value_counts())

'Value counts for column: SID'

SID
N2P7GLCC6    1
UHDOP5522    1
UHE842CU6    1
UHJFT1JAB    1
UHKF05TAF    1
            ..
UIEFJU8PD    1
UI9P657PI    1
UI98YGQUI    1
UI8PIQL96    1
UI3BNGXIW    1
Name: count, Length: 25198, dtype: int64

'Value counts for column: COHORT'

COHORT
Fall 2022    5362
Fall 2019    5158
Fall 2018    4909
Fall 2020    4904
Fall 2021    4865
Name: count, dtype: int64

'Value counts for column: RACE_ETHNICITY'

RACE_ETHNICITY
Hispanic                                     12327
Asian                                         5987
White                                         3467
Two or More Races                             1186
Nonresident alien                              922
Black or African American                      910
Unknown                                        317
Native Hawaiian or Other Pacific Islander       60
American Indian or Alaska Native                22
Name: count, dtype: int64

'Value counts for column: GENDER'

GENDER
Female       15072
Male          9996
Nonbinary       38
female          33
 Male           24
 Female         22
male            13
Name: count, dtype: int64

'Value counts for column: FIRST_GEN_STATUS'

FIRST_GEN_STATUS
Continuing Generation    15689
First Generation          7368
Unknown                   2141
Name: count, dtype: int64

'Value counts for column: COLLEGE'

COLLEGE
Health & Human Services              4561
Engineering & Technology             4041
General Studies                      3943
Letters & Humanities                 3692
Natural and Mathematical Sciences    2864
Business Administration              2834
Visual & Performing Arts             2752
Education & Leadership                511
Name: count, dtype: int64

'Value counts for column: SEM_1_STATUS'

SEM_1_STATUS
E    25198
Name: count, dtype: int64

'Value counts for column: SEM_2_STATUS'

SEM_2_STATUS
E    24047
D     1151
Name: count, dtype: int64

'Value counts for column: SEM_3_STATUS'

SEM_3_STATUS
E    21738
D     3460
Name: count, dtype: int64

Fix inconsistent labels in the 'GENDER' column by converting all entries to a consistent case and removing leading/trailing spaces.



In [25]:
df['GENDER'] = df['GENDER'].str.strip().str.capitalize()
display(df['GENDER'].value_counts())

GENDER
Female       15127
Male         10033
Nonbinary       38
Name: count, dtype: int64

Combine or eliminate rare categories.

In [26]:
# Eliminate non-binary category
df = df[df['GENDER'] != 'Non-binary']

# For RACE_ETHNICITY compbine Unknown, Native Hawaiian or Other Pacific Islander, and American Indian or Alaska Native into the category Other
df['RACE_ETHNICITY'] = df['RACE_ETHNICITY'].replace(['Unknown', 'Native Hawaiian or Other Pacific Islander', 'American Indian or Alaska Native'], 'Other')

## Data splitting

Split the cleaned data into training and testing dataframes.


In [27]:
from sklearn.model_selection import train_test_split

df_train, df_test = train_test_split(df, test_size=0.25, random_state=42)

display(f"Shape of training data: {df_train.shape}")
display(f"Shape of testing data: {df_test.shape}")

'Shape of training data: (18898, 18)'

'Shape of testing data: (6300, 18)'

Cohort based split

In [28]:
# Create the test set with the 'Fall 2022' cohort
df_test_cohort = df[df['COHORT'] == 'Fall 2022'].copy()

# Create the training set with the remaining cohorts
df_train_cohort = df[df['COHORT'] != 'Fall 2022'].copy()

display(f"Shape of training data (cohort split): {df_train_cohort.shape}")
display(f"Shape of testing data (cohort split): {df_test_cohort.shape}")

'Shape of training data (cohort split): (19836, 18)'

'Shape of testing data (cohort split): (5362, 18)'

Method 1: Random Split (train_test_split)

    Pros:
        Ensures that both the training and testing sets are representative of the overall dataset's distribution. This is crucial for building a model that generalizes well to unseen data that comes from the same distribution as the training data.
        Simple to implement and a standard practice in machine learning.
        Avoids potential biases that could arise from non-random splits.

    Cons:
        May not be ideal if you need to evaluate your model's performance on a specific, future cohort or a group with unique characteristics.
        If there are significant differences between cohorts, a randomly split test set might not accurately reflect how the model would perform on a completely new, future cohort.

Method 2: Cohort-Based Split (COHORT == 'Fall 2022')

    Pros:
        Provides a realistic evaluation of how your model would perform on a specific group, such as a future cohort. This is particularly useful in time-series or sequential data where you want to predict outcomes for a future period based on past data.
        Allows you to assess the model's ability to generalize to a cohort that may have different characteristics or trends compared to the training cohorts.

    Cons:
        The test set may not be representative of the overall data distribution if the chosen cohort has unique characteristics. This could lead to an over- or underestimation of the model's performance.
        If the chosen test cohort is significantly different from the training cohorts, the model trained on the earlier cohorts might not perform well due to concept drift or changes in underlying patterns.
        Reduces the size of the training data, which could impact model performance, especially for complex models that require a large amount of data.

When to use which method:

    Use random splitting when you want to build a model that generalizes well to new data from the same distribution as your training data and you don't have specific concerns about evaluating performance on a distinct group or time period.
    Use cohort-based splitting when you need to specifically evaluate your model's performance on a future or distinct group, such as the most recent cohort, to understand how it would perform in a real-world scenario with new data.

In our case, using the 'Fall 2022' cohort as a test set might be a good approach if our goal is to predict the retention of the most recent cohort based on data from previous cohorts. This simulates a real-world scenario where you would use historical data to predict outcomes for new students. However, it's important to be aware of the potential limitations, such as the representativeness of the 'Fall 2022' cohort compared to the others.

Impute missing values in both df_train and df_test separately.


Identify columns with missing values in df_train and df_test and impute them separately using medians for numerical columns and modes for categorical columns, calculating these statistics from df_train only for both dataframes, and finally verify that there are no remaining missing values.



In [29]:
def impute_missing_values(df_train, df_test):
  # Identify columns with missing values in df_train
  missing_train = df_train.isnull().sum()
  cols_with_missing_train = missing_train[missing_train > 0].index

  # Impute missing values in df_train
  for col in cols_with_missing_train:
      if df_train[col].dtype in ['int64', 'float64']:
          median_val = df_train[col].median()
          df_train[col].fillna(median_val, inplace=True)
      else:
          mode_val = df_train[col].mode()[0]
          df_train[col].fillna(mode_val, inplace=True)

  # Identify columns with missing values in df_test (based on original df columns)
  missing_test = df_test.isnull().sum()
  cols_with_missing_test = missing_test[missing_test > 0].index

  # Impute missing values in df_test using statistics from df_train
  for col in cols_with_missing_test:
      if df_test[col].dtype in ['int64', 'float64']:
          # Use median from df_train
          median_val_train = df_train[col].median()
          df_test[col].fillna(median_val_train, inplace=True)
      else:
          # Use mode from df_train
          mode_val_train = df_train[col].mode()[0]
          df_test[col].fillna(mode_val_train, inplace=True)


  return df_train, df_test

df_train, df_test = impute_missing_values(df_train, df_test)
df_train_cohort, df_test_cohort = impute_missing_values(df_train_cohort, df_test_cohort)

<ipython-input-29-1579737130>:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_train[col].fillna(median_val, inplace=True)
<ipython-input-29-1579737130>:24: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'd

In [30]:
# Verify no remaining missing values
display("Missing values in df_train after imputation:")
display(df_train_cohort.isnull().sum().sum())
display("Missing values in df_test after imputation:")
display(df_test_cohort.isnull().sum().sum())

'Missing values in df_train after imputation:'

np.int64(0)

'Missing values in df_test after imputation:'

np.int64(0)

## Summary:

### Data Analysis Key Findings

*   The initial dataset contained 25266 rows and 19 columns.
*   54 duplicate rows were found and removed from the dataset.
*   The 'FAMILY\_INCOME' column had a significant number of missing values (20920) and was subsequently dropped as it exceeded the 50% missing value threshold.
*   The 'GENDER' column contained inconsistent casing and leading/trailing spaces which were standardized.
*   The dataset was successfully split into training (75%) and testing (25%) sets, resulting in `df_train` and `df_test` .
*   The dataset was successfully split into training (cohorts from Fall 2018 through Fall 2022) and testing (cohort of Fall 2023) sets, resulting in `df_train_cohort and `df_test_cohort`.
*   Missing values in both the training and testing sets were imputed using the median for numerical columns and the mode for categorical columns, with the imputation statistics calculated solely from the training data to prevent data leakage.

### Insights or Next Steps

*   The preprocessing steps have prepared the data for model training by handling duplicates, removing columns with excessive missing data, splitting the data, and imputing missing values in a leakage-aware manner.
*   The next step would involve further feature engineering and selecting appropriate machine learning models for the retention prediction task.


In [31]:
# Save df_train_cohort to a CSV file
df_train_cohort.to_csv(f'{location}training.csv', index=False)

# Save df_test_cohort to a CSV file
df_test_cohort.to_csv(f'{location}testing.csv', index=False)